spring-kafka/src/main/java/com/baeldung/partitioningstrategy/KafkaMessageConsumer.java

```diff
@@ -0,0 +1,30 @@
+package com.baeldung.partitioningstrategy;
+
+import java.util.ArrayList;
+import java.util.List;
+import org.springframework.kafka.annotation.KafkaListener;
+import org.springframework.messaging.handler.annotation.Header;
+import org.springframework.messaging.handler.annotation.Payload;
+import org.springframework.stereotype.Service;
+import org.springframework.kafka.support.KafkaHeaders;
+import jakarta.annotation.Nullable;
+
+@Service
+public class KafkaMessageConsumer {
+    private final List<ReceivedMessage> receivedMessages = new ArrayList<>();
```

Comment:
This is not thread safe, can we use a CopyOnWriteArrayList or similar instead?

Or, do you know if @KafkaListener is guaranteed to be single threaded?

Reply:
I've checked, the @KafkaListener is not inherently thread-safe. Thanks for pointing out this, will change to CopyOnWriteArrayList instead.